src/Microsoft.Health.Fhir.Core/Features/Storage/FhirMemoryCache.cs

```diff
@@ -0,0 +1,148 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Runtime.Caching;
+using EnsureThat;
+
+namespace Microsoft.Health.Fhir.Core.Features.Storage
+{
+    public sealed class FhirMemoryCache<T>
+    {
+        private const int DefaultLimitSizeInMegabytes = 50;
+        private const int DefaultExpirationTimeInMinutes = 60;
+
+        private readonly ObjectCache _cache = MemoryCache.Default;
+        private readonly TimeSpan _expirationTime;
+        private readonly object _lock;
+
+        public FhirMemoryCache(string name)
+            : this(
+                  name,
+                  limitSizeInMegabytes: DefaultLimitSizeInMegabytes,
+                  expirationTime: TimeSpan.FromMinutes(DefaultExpirationTimeInMinutes))
+        {
+        }
+
+        public FhirMemoryCache(string name, int limitSizeInMegabytes, TimeSpan expirationTime)
+        {
+            EnsureArg.IsNotNull(name, nameof(name));
+            EnsureArg.IsGt(limitSizeInMegabytes, 0, nameof(name));
+
+            _cache = new MemoryCache(
+                name,
+                new NameValueCollection()
+                {
+                    { "CacheMemoryLimitMegabytes", limitSizeInMegabytes.ToString() },
+                });
+
+            _expirationTime = expirationTime;
```

Comment:
Is there an LRU type option? Would LRU or a sliding expiration be better?

Reply:
As discussed, we are implementing the sliding window caching.